File: go-leetcode/p0206_reverse-linked-list.go
package leetcode

// https://leetcode.com/problems/reverse-linked-list/description/

func reverseList(head *IntListNode) *IntListNode {
	var prev *IntListNode

	cur := head
	for cur != nil {
		next := cur.Next
		cur.Next = prev
		prev, cur = cur, next
	}

	return prev
}

File: rust-leetcode/src/p0206_reverse_linked_list.rs
// https://leetcode.com/problems/reverse-linked-list/

pub struct Solution;

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }

// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }

impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut prev = None;
        let mut cur = head;

        while let Some(mut node) = cur {
            let next = node.next;
            node.next = prev;
            prev = Some(node);
            cur = next;
        }

        prev
    }
}
